src/hotel_reservations/utils.py

```diff
@@ -149,52 +156,116 @@ def generate_booking_ids_regex(existing_ids: list[str], num_new_ids=1000, prefix
+            StructField("lead_time", IntegerType(), True),
+            StructField("arrival_year", IntegerType(), True),
+            StructField("arrival_month", IntegerType(), True),
+            StructField("arrival_date", IntegerType(), True),
+            StructField("market_segment_type", StringType(), True),
+            StructField("repeated_guest", IntegerType(), True),
+            StructField("no_of_previous_cancellations", IntegerType(), True),
+            StructField("no_of_previous_bookings_not_canceled", IntegerType(), True),
+            StructField("avg_price_per_room", DoubleType(), True),
+            StructField("no_of_special_requests", IntegerType(), True),
+            StructField("booking_status", StringType(), True),
         ]
-        for key, feature in config.cat_features.items()
-    }
-    for col_name, allowed_values in cat_features.items():
-        synthetic_data[col_name] = np.random.choice(allowed_values, num_rows)
+    )
 
-    # Create target variable (booking_status) as a random value
-    synthetic_data[config.target] = np.random.choice(["Not_Canceled", "Canceled"], num_rows, p=[0.9, 0.1])
-
-    existing_ids = input_data.select(config.primary_key).rdd.flatMap(lambda x: x).collect()
+    # Extract existing booking IDs
+    existing_ids = input_data.select("Booking_ID").rdd.flatMap(lambda x: x).collect()
     synthetic_ids = generate_booking_ids_regex(existing_ids, num_new_ids=num_rows)
-    synthetic_data[config.primary_key] = synthetic_ids
 
-    # Convert the synthetic data dictionary to a DataFrame
-    synthetic_df = spark.createDataFrame(pd.DataFrame(synthetic_data))
+    # Generate synthetic data
+    def generate_row(booking_id):
+        return {
+            "Booking_ID": booking_id,
+            "no_of_adults": random.randint(1, 4),
+            "no_of_children": random.randint(0, 3),
+            "no_of_weekend_nights": random.randint(0, 3),
+            "no_of_week_nights": random.randint(0, 7),
+            "type_of_meal_plan": random.choice(["Meal_Plan_1", "Meal_Plan_2", "Meal_Plan_3", "Not Selected"]),
+            "required_car_parking_space": random.randint(0, 1),
+            "room_type_reserved": random.choice(["Room_Type_1", "Room_Type_2", "Room_Type_3", "Room_Type_4"]),
+            "lead_time": random.randint(0, 365),
+            "arrival_year": 2024,
+            "arrival_month": random.randint(1, 12),
+            "arrival_date": random.randint(1, 28),
+            "market_segment_type": random.choice(["Online", "Corporate", "Offline"]),
+            "repeated_guest": random.randint(0, 1),
+            "no_of_previous_cancellations": random.randint(0, 2),
+            "no_of_previous_bookings_not_canceled": random.randint(0, 5),
+            "avg_price_per_room": round(random.uniform(50.0, 300.0), 2),
+            "no_of_special_requests": random.randint(0, 3),
+            "booking_status": random.choice(["Canceled", "Not_Canceled"]),
+        }
```

Comment:
Since the synthetic data in this use case is only used for demonstrational purposes in this assignment (and not for actual customer serving pipelines), I will leave it as it is.